ide/rust/SyncRunner.kt

```diff
@@ -0,0 +1,115 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.vaticle.dependencies.ide.rust
+
+import com.vaticle.bazel.distribution.common.Logging.LogLevel.DEBUG
+import com.vaticle.bazel.distribution.common.Logging.LogLevel.ERROR
+import com.vaticle.bazel.distribution.common.Logging.Logger
+import com.vaticle.bazel.distribution.common.shell.Shell
+import com.vaticle.bazel.distribution.common.util.FileUtil.listFilesRecursively
+import com.vaticle.dependencies.ide.rust.SyncRunner.ShellArgs.BAZEL
+import com.vaticle.dependencies.ide.rust.SyncRunner.ShellArgs.BUILD
+import com.vaticle.dependencies.ide.rust.SyncRunner.ShellArgs.INFO
+import com.vaticle.dependencies.ide.rust.SyncRunner.ShellArgs.QUERY
+import picocli.CommandLine
+import java.io.File
+import java.nio.file.Path
+import java.util.concurrent.Callable
+import kotlin.io.path.Path
+import kotlin.system.exitProcess
+
+fun main(args: Array<String>): Unit = exitProcess(CommandLine(SyncRunner()).execute(*args))
+
+@CommandLine.Command(name = "sync", mixinStandardHelpOptions = true)
+class SyncRunner : Callable<Unit> {
+
+    @CommandLine.Option(names = ["--verbose", "-v"], required = false)
+    var verbose: Boolean = false
+
+    lateinit var logger: Logger
+    private lateinit var shell: Shell
+    private lateinit var buildWorkspaceDir: Path
+    lateinit var bazelOutputBase: Path
+
+    override fun call() {
+        logger = Logger(logLevel = if (verbose) DEBUG else ERROR)
+        shell = Shell(logger, verbose)
+        buildWorkspaceDir = Path(System.getenv("BUILD_WORKSPACE_DIRECTORY"))
+        bazelOutputBase = Path(shell.execute(listOf(BAZEL, INFO, "output_base"), buildWorkspaceDir).outputString().trim())
+        val rustTargets = rustTargets(buildWorkspaceDir)
+        validateBuildWorkspace(rustTargets)
+        loadRustToolchainAndExternalDeps(rustTargets)
+        rustRepoPaths().forEach { sync(it) }
+    }
+
+    private fun validateBuildWorkspace(rustTargets: List<String>) {
+        shell.execute(command = listOf(BAZEL, BUILD) + rustTargets + "--build=false", baseDir = buildWorkspaceDir)
+    }
+
+    private fun loadRustToolchainAndExternalDeps(rustTargets: List<String>) {
+        shell.execute(command = listOf(BAZEL, BUILD) + rustTargets, baseDir = buildWorkspaceDir, throwOnError = false)
+    }
+
+    private fun rustTargets(repoPath: Path): List<String> {
+        return shell.execute(listOf(BAZEL, QUERY, "kind(rust_*, //...)"), repoPath)
+            .outputString().split("\n").filter { it.isNotBlank() }
+    }
+
+    private fun rustRepoPaths(): List<Path> {
+        // e.g: [/Users/root/workspace/typedb-client-rust, /private/var/_bazel_root_/123abc/external/vaticle_typedb_protocol]
+        return listOf(buildWorkspaceDir) + shell.execute(listOf(BAZEL, QUERY, "kind(rust_*, deps(//...))", "--output=package"), buildWorkspaceDir)
+            .outputString().split("\n")
+            .filter { it.startsWith("@vaticle_") }
```

Comment:
Paths like these will break when we refactor the repo. Is there any way that we can test them? At the very least, we gather them into a single "constant" or "config" object.